### PR TITLE
UI: use YUV stream instead of RGB

### PR DESCRIPTION
**Pull request by jwooning**
redo of #24298 but including memcpy, removing that will be a separate PR

### CODE REVIEW

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 1]
```diff
@@ -209,13 +224,17 @@ void CameraViewWidget::paintGL() {
   }
 
   glBindVertexArray(frame_vao);
-  glActiveTexture(GL_TEXTURE0);
-  glBindTexture(GL_TEXTURE_2D, texture[latest_texture_id]->frame_tex);
 
   glUseProgram(program->programId());
-  glUniform1i(program->uniformLocation("uTexture"), 0);
-  glUniformMatrix4fv(program->uniformLocation("uTransform"), 1, GL_TRUE, frame_mat.v);
+  for (int i = 0; i < 3; ++i) {
```
Comment (geohot) [inline selfdrive/ui/qt/widgets/cameraview.cc:229]: I don't understand why this loop has to be here at all.

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 2]
```diff
@@ -290,19 +310,30 @@ void CameraViewWidget::vipcThread() {
...
+          void *texture_buffer = gl_buffer->map(QOpenGLBuffer::WriteOnly);
+
+          if (texture_buffer == nullptr) {
+            LOGE("gl_buffer->map returned nullptr");
+            continue;
+          }
+
+          int width = i == 0 ? stream_width : stream_width / 2;
+          int height = i == 0 ? stream_height : stream_height / 2;
+          if (i == 0) {
```
[the remark this answers is not in the document]
Comment (geohot) [reply selfdrive/ui/qt/widgets/cameraview.cc:323]: General C++ question, is there a way to do that select on one line without that nested if thing? Like I'm playing with:
```
char *tmp[] = {argv[2], argv[3], argv[4]};
char *s = tmp[atoi(argv[1])];
```

File: selfdrive/ui/qt/widgets/cameraview.cc (hunk 1 above)
Comment (jwooning) [reply selfdrive/ui/qt/widgets/cameraview.cc:229]: It tells the shader which texture unit points to which which texture buffer

File: selfdrive/ui/qt/widgets/cameraview.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (pd0wm) [reply selfdrive/ui/qt/widgets/cameraview.cc]:
```c++
          uint8_t* tex_buf[] = {buf->y, buf->u, buf->v};
          memcpy(texture_buffer, tex_buf[i], width*height);
```
seems to work fine

**Comment by zorrobyte** [comment]
@jwooning this breaks replay and webcam, see above issue ^^

**Comment by zorrobyte** [comment]
> Does #24418 fix it for you?

It does, thank you!

**Pull request merged**